### PR TITLE
Proof of Concept: Frame Timing Integration

### PR DESCRIPTION
**Pull request by ueman**
## :scroll: Description
Proof of concept for tracking frame durations.

It's not really useful without being able to better determine what's causing it.

## :bulb: Motivation and Context
#513

## :green_heart: How did you test it?
Not yet

## :pencil: Checklist

- [ ] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] I updated the docs if needed
- [ ] All tests passing
- [ ] No breaking changes

### CODE REVIEW

File: flutter/lib/src/integrations/frame_timing_integration.dart [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+/// See [SchedulerBinding.addTimingsCallback](https://api.flutter.dev/flutter/scheduler/SchedulerBinding/addTimingsCallback.html)
+/// to learn more about the performance impact of using this.
+///
+/// Adding a `timingsCallback` has a real significant performance impact as
+/// noted above. Thus this integration should only be added if it's enabled.
+/// The enabled check should not happen inside the `timingsCallback`.
+class FrameTimingIntegration extends Integration<SentryFlutterOptions> {
+  FrameTimingIntegration({
+    required this.reporter,
+    this.badFrameThreshold = const Duration(milliseconds: 16),
```
Comment (ueman) [inline flutter/lib/src/integrations/frame_timing_integration.dart:25]: I haven't yet found a way to determine the device current refresh rate.

File: flutter/lib/src/integrations/frame_timing_integration.dart [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+      level: SentryLevel.warning,
+    ));
+  }
+
+  String _message(
+    int badFrameCount,
+    Duration worstFrameDuration,
+    Duration totalDuration,
+  ) {
+    return '$badFrameCount frames exceeded ${_formatMS(badFrameThreshold)} '
```
Comment (ueman) [inline flutter/lib/src/integrations/frame_timing_integration.dart:99]: Average frame time would be nice to know, too.

**Comment by ueman** [comment]
Now that we have performance tracing, we can add frame metrics just like [Android](https://github.com/getsentry/sentry-java/blob/1a409209b77d5e28915c2f3df49e6941bed3d79c/sentry-android-core/src/main/java/io/sentry/android/core/ActivityFramesTracker.java#L96-L98) or [iOS](https://github.com/getsentry/sentry-cocoa/blob/94ec24f58b12c5c7dfae7d9f6430c06655026201/Sources/Sentry/SentryFramesTracker.m)

File: flutter/example/lib/main.dart [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import 'dart:async';
+import 'dart:io';
```
Comment (marandaneto) [inline flutter/example/lib/main.dart:2]: that requires a custom import since it should run on Web too

**Comment by ueman** [comment]
@marandaneto Is there some documentation on how these values should be added to transactions? Or do I have to go by the Android/iOS source code?

**Comment by marandaneto** [comment]
> @marandaneto Is there some documentation on how these values should be added to transactions? Or do I have to go by the Android/iOS source code?

there are no official docs yet since mobile vitals is a preview, but if you're willing to work on that, I can guide you.

**Comment by ueman** [comment]
@marandaneto that would be nice

**Comment by ueman** [comment]
If I understand the source code of the Android SDK correctly, I need to add the frame metrics in the following format to a transaction?
```json
{
...
"measurements":{
      "frames_frozen":{
         "value":0.0
      },
      "frames_slow":{
         "value":2.0
      },
      "frames_total":{
         "value":2.0
      }
   }
},
...
```
I'm assuming the data collection should start (and end) with the start (and end) of a transaction. Is that correct?

**Comment by marandaneto** [comment]
> If I understand the source code of the Android SDK correctly, I need to add the frame metrics in the following format to a transaction?
> 
> ```json
> {
> ...
> "measurements":{
>       "frames_frozen":{
>          "value":0.0
>       },
>       "frames_slow":{
>          "value":2.0
>       },
>       "frames_total":{
>          "value":2.0
>       }
>    }
> },
> ...
> ```
> 
> I'm assuming the data collection should start (and end) with the start (and end) of a transaction. Is that correct?

that's correct, good finding, sorry, we didn't have the time to document it yet.

**Comment by ueman** [comment]
Are there any hooks for a transaction start and end which I can use to turn on and off frame time recording? The docs state that it has a small performance impact, so I really don't want to have it enabled the whole time.

If there's no hooks, I can think of three ways to enable us to go further with this:
- Add some kind of hooks for transactions. Maybe something like integrations/event processors but just internal?
- A transaction factory which an SDK can register. The Flutter SDK could then subclass the transaction to add frame time recordings
- A method on `SentryFlutter` to start a subclassed transaction which enables frame time recordings.

**Comment by marandaneto** [comment]
> Are there any hooks for a transaction start and end which I can use to turn on and off frame time recording? The docs state that it has a small performance impact, so I really don't want to have it enabled the whole time.
> 
> If there's no hooks, I can think of three ways to enable us to go further with this:
> 
> * Add some kind of hooks for transactions. Maybe something like integrations/event processors but just internal?
> * A transaction factory which an SDK can register. The Flutter SDK could then subclass the transaction to add frame time recordings
> * A method on `SentryFlutter` to start a subclassed transaction which enables frame time recordings.

ideally, like Android, the measurements are dependent on this issue https://github.com/getsentry/sentry-dart/pull/611
so we start checking slow/frozen frames when a transaction is created by the navigation transactions.
and then thru the navigation transaction, we can start and stop the measurements monitoring, so no need for hooks or anything, just an integration that handles everything, well, we do need hooks to start and finish a transaction during the screen loading (or routes).

**Pull request closed without merging**